File: apps/go/chapter05/recursion_test.go
package chapter05_test

import (
	"testing"

	"github.com/k2works/getting-started-algorithm/go/chapter05"
)

func TestFactorial(t *testing.T) {
	got := chapter05.Factorial(5)
	if got != 120 {
		t.Errorf("Factorial(5) = %d; want 120", got)
	}
}

func TestGcd(t *testing.T) {
	got := chapter05.Gcd(22, 8)
	if got != 2 {
		t.Errorf("Gcd(22,8) = %d; want 2", got)
	}
}

func TestRecursiveSum(t *testing.T) {
	got := chapter05.RecursiveSum(5)
	if got != 15 {
		t.Errorf("RecursiveSum(5) = %d; want 15", got)
	}
}

func TestHanoi(t *testing.T) {
	moves := chapter05.Hanoi(1, "A", "C", "B")
	if len(moves) != 1 || moves[0][0] != "A" || moves[0][1] != "C" {
		t.Errorf("Hanoi(1,A,C,B) = %v; want [[A C]]", moves)
	}

	moves3 := chapter05.Hanoi(3, "A", "C", "B")
	if len(moves3) != 7 {
		t.Errorf("Hanoi(3,...) len = %d; want 7", len(moves3))
	}
}

func TestMazeSolve(t *testing.T) {
	maze := [][]int{
		{1, 1, 1},
		{1, 0, 1},
		{1, 1, 1},
	}
	got := chapter05.MazeSolve(maze, 1, 1, 1, 1, nil)
	if !got {
		t.Error("MazeSolve should return true for start==goal")
	}
}

File: apps/c/chapter05/recursion_test.c
#include <stdio.h>
#include "../include/test_helper.h"
#include "recursion.h"

void test_factorial(void) {
    ASSERT_EQ_INT(120, factorial(5));
    ASSERT_EQ_INT(1,   factorial(0));
}

void test_gcd(void) {
    ASSERT_EQ_INT(2, gcd(22, 8));
    ASSERT_EQ_INT(6, gcd(18, 24));
}

void test_recursive_sum(void) {
    ASSERT_EQ_INT(15, recursive_sum(5));
    ASSERT_EQ_INT(0,  recursive_sum(0));
}

void test_hanoi(void) {
    HanoiMove moves[10];
    int n = hanoi(1, 'A', 'C', 'B', moves);
    ASSERT_EQ_INT(1, n);
    ASSERT_EQ_INT('A', moves[0].from);
    ASSERT_EQ_INT('C', moves[0].to);

    n = hanoi(3, 'A', 'C', 'B', moves);
    ASSERT_EQ_INT(7, n);
}

void test_maze_solve(void) {
    int maze[3][3] = {{1,1,1},{1,0,1},{1,1,1}};
    ASSERT_TRUE(maze_solve((int *)maze, 3, 3, 1, 1, 1, 1));
}

void test_eight_queen(void) {
    int result_count = eight_queen();
    ASSERT_EQ_INT(8*8*8*8*8*8*8*8, result_count); /* 8^8 = 16777216 */
}

void test_eight_queen2(void) {
    int result_count = eight_queen2();
    ASSERT_EQ_INT(40320, result_count); /* 8! = 40320 */
}

void test_eight_queen3(void) {
    int result_count = eight_queen3();
    ASSERT_EQ_INT(92, result_count);
}

int main(void) {
    test_factorial();
    test_gcd();
    test_recursive_sum();
    test_hanoi();
    test_maze_solve();
    test_eight_queen();
    test_eight_queen2();
    test_eight_queen3();
    TEST_SUMMARY();
    return 0;
}
